test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineVerifyError.java

```diff
@@ -0,0 +1,100 @@
+ */
+
+/*
+ * @test
+ * @bug 6402717 8330606
+ * @summary Redefine VerifyError to get a VerifyError should not throw SOE
+ * @requires vm.jvmti
+ * @library /test/lib
+ * @modules java.base/jdk.internal.misc
+ *          java.base/jdk.internal.org.objectweb.asm
+ *          java.compiler
+ *          java.instrument
+ *          jdk.jartool/sun.tools.jar
+ * @run main RedefineClassHelper
+ * @run main/othervm/timeout=180
+ *         -javaagent:redefineagent.jar
+ *         -Xlog:class+init,exceptions
+ *         RedefineVerifyError
+ */
+
+import jdk.internal.org.objectweb.asm.AnnotationVisitor;
+import jdk.internal.org.objectweb.asm.Attribute;
+import jdk.internal.org.objectweb.asm.ClassReader;
+import jdk.internal.org.objectweb.asm.ClassWriter;
+import jdk.internal.org.objectweb.asm.ConstantDynamic;
+import jdk.internal.org.objectweb.asm.FieldVisitor;
+import jdk.internal.org.objectweb.asm.Handle;
+import jdk.internal.org.objectweb.asm.Label;
+import jdk.internal.org.objectweb.asm.MethodVisitor;
+import jdk.internal.org.objectweb.asm.Opcodes;
+import jdk.internal.org.objectweb.asm.RecordComponentVisitor;
+import jdk.internal.org.objectweb.asm.Type;
+import jdk.internal.org.objectweb.asm.TypePath;
+
+public class RedefineVerifyError implements Opcodes {
+
+    // This is a redefinition of java.lang.VerifyError with two broken init methods (no bytecodes)
+    public static byte[] dump () throws Exception {
+
+        ClassWriter classWriter = new ClassWriter(0);
+        FieldVisitor fieldVisitor;
+        RecordComponentVisitor recordComponentVisitor;
+        MethodVisitor methodVisitor;
+        AnnotationVisitor annotationVisitor0;
+
+        classWriter.visit(52, ACC_SUPER | ACC_PUBLIC, "java/lang/VerifyError", null, "java/lang/LinkageError", null);
+        {
+            fieldVisitor = classWriter.visitField(ACC_PRIVATE | ACC_FINAL | ACC_STATIC, "serialVersionUID", "J", null, new Long(7001962396098498785L));
+            fieldVisitor.visitEnd();
+        }
+        {
+            methodVisitor = classWriter.visitMethod(ACC_PUBLIC, "<init>", "()V", null, null);
+            methodVisitor.visitCode();
+            methodVisitor.visitEnd();
+        }
+
+        {
+            methodVisitor = classWriter.visitMethod(ACC_PUBLIC, "<init>", "(Ljava/lang/String;)V", null, null);
+            methodVisitor.visitCode();
+            classWriter.visitEnd();
+        }
+
+        return classWriter.toByteArray();
+    }
+
+    public static void main(String[] args) throws Exception {
+
+        Class<?> verifyErrorMirror = java.lang.VerifyError.class;
+
+        try {
+            // The Verifier is called for the redefinition, which will fail because of the broken <init> method above.
+            RedefineClassHelper.redefineClass(verifyErrorMirror, dump());
+            throw new RuntimeException("This should throw VerifyError");
+        } catch (VerifyError e) {
+            // JVMTI recreates the VerifyError so the verification message is lost.
```

Comment:
I'm not clear why JVMTI would mess with the `VerifyError` that the verifier should have raised, and which explains the reason why verification failed. Not that your changes impact that.

Reply:
The JVMTI code for redefinition returns JVMTI_ERROR_FAILS_VERIFICATION and not the pending exception to the caller.  The agent eventually in java.instrument/share/native/libinstrument/JavaExceptions.c will recreate the VerifyError based on the JVMTI error code.

Reply:
Wow that is pretty developer unfriendly. Hopefully exception logging will still show what the error is in detail.

Reply:
Yes, logging does show the detailed verify error. The test has -Xlog:exceptions also that shows where it came from.